File: src/lib/types/message.ts
export type Message =
  | UserMessage
  | AssistantMessage
  | ReasoningStep
  | ToolApprovalRequest
  | ToolApprovalResponse
  | ToolExecution
  | SystemMessage
  | ErrorMessage;

export interface UserMessage {
  type: 'UserMessage';
  content: string;
}

export interface AssistantMessage {
  type: 'AssistantMessage';
  content: string;
  model: string;
}

export interface ReasoningStep {
  type: 'ReasoningStep';
  step: string;
  model: string;
  content: string;
}

export interface ToolApprovalRequest {
  type: 'ToolApprovalRequest';
  request_id: string;
  tool_name: string;
  parameters: Record<string, any>;
  reasoning: string;
}

export interface ToolApprovalResponse {
  type: 'ToolApprovalResponse';
  request_id: string;
  approved: boolean;
}

export interface ToolExecution {
  type: 'ToolExecution';
  tool_name: string;
  parameters: Record<string, any>;
  result: any;
}

export interface SystemMessage {
  type: 'SystemMessage';
  content: string;
}

export interface ErrorMessage {
  type: 'Error';
  message: string;
}

File: src-tauri/src/models/message.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum Message {
    UserMessage {
        content: String,
    },
    AssistantMessage {
        content: String,
        model: String,
    },
    ReasoningStep {
        step: String,
        model: String,
        content: String,
    },
    ToolApprovalRequest {
        request_id: String,
        tool_name: String,
        parameters: serde_json::Value,
        reasoning: String,
    },
    ToolApprovalResponse {
        request_id: String,
        approved: bool,
    },
    ToolExecution {
        tool_name: String,
        parameters: serde_json::Value,
        result: serde_json::Value,
    },
    SystemMessage {
        content: String,
    },
    Error {
        message: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SidecarRequest {
    pub session_id: String,
    pub message: Message,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SidecarResponse {
    pub success: bool,
    pub data: Option<serde_json::Value>,
    pub error: Option<String>,
}
